vm-virtio/src/vhost_user/blk.rs

```diff
@@ -0,0 +1,332 @@
+    num_queues: u16,
+    max_discard_sectors: u32,
+    max_discard_seg: u32,
+    discard_sector_alignment: u32,
+    max_write_zeroes_sectors: u32,
+    max_write_zeroes_seg: u32,
+    write_zeroes_may_unmap: u8,
+    unused1: [u8; 3],
+}
+
+pub struct Blk {
+    vhost_user_blk: Master,
+    kill_evt: EventFd,
+    avail_features: u64,
+    acked_features: u64,
+    config_space: Vec<u8>,
+    queue_sizes: Vec<u16>,
+}
+
+impl Blk {
+    /// Create a new vhost-user-blk device
+    pub fn new(path: &str, num_queues: usize, queue_size: u16, config_wce: u8) -> Result<Blk> {
+        // Connect to the vhost-user socket.
+        let mut vhost_user_blk =
+            Master::connect(path, num_queues as u64).map_err(Error::VhostUserCreateMaster)?;
+
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateKillEventFd)?;
+        // Retrieve available features only when connecting the first time.
+        let mut avail_features = vhost_user_blk.get_features().map_err(Error::VhostUserGetFeatures)?;
+        // Let only ack features we expect, that is VIRTIO_F_VERSION_1.
+        if (avail_features & VIRTIO_F_VERSION_1_BITMASK) != VIRTIO_F_VERSION_1_BITMASK {
+            return Err(Error::InvalidFeatures);
+        }
+        avail_features =
+            VIRTIO_F_VERSION_1_BITMASK | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
+        vhost_user_blk
+            .set_features(avail_features)
+            .map_err(Error::VhostUserSetFeatures)?;
+        // Identify if protocol features are supported by the slave.
+        if (avail_features & VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits())
+            == VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+        {
+            let mut protocol_features = vhost_user_blk
+                .get_protocol_features()
+                .map_err(Error::VhostUserGetProtocolFeatures)?;
+            protocol_features |= VhostUserProtocolFeatures::MQ;
+            vhost_user_blk
+                .set_protocol_features(protocol_features)
+                .map_err(Error::VhostUserSetProtocolFeatures)?;
+        }
+
+        vhost_user_blk
+            .set_owner()
+            .map_err(Error::VhostUserSetOwner)?;
+
+        let config_len = mem::size_of::<virtio_blk_config>();
+        let mut config_space = Vec::with_capacity(config_len as usize);
+        config_space.resize(config_len as usize, 0);
+
+        let offset = offset_of!(virtio_blk_config, wce);
+        // only set wce value.
+        config_space[offset] = config_wce;
+
+        let queue_num_offset = offset_of!(virtio_blk_config, num_queues);
+        // only setnum_queues value.
+        config_space[queue_num_offset] = num_queues as u8;
+
+        Ok(Blk {
+            vhost_user_blk: vhost_user_blk,
+            kill_evt,
+            avail_features,
+            acked_features: 0u64,
+            config_space,
+            queue_sizes: vec![queue_size; num_queues],
+        })
+    }
+
+    pub fn setup_vub(
```

Comment:
Same here, this is generic to vhost-user protocol and I would expect this to be factorized between all vhost-user implementations (fs, net, and blk).